cgi-bin/LJ/User.pm

```diff
@@ -3195,61 +3198,121 @@ sub thread_expand_all {
+        my %sticky_hash = map { $_ =>  1 }  @input;
+        foreach my $unused_sticky ( @currently_unused_stickies ) {
+            push @new_unused_stickies, $unused_sticky unless exists $sticky_hash{$unused_sticky};
+        }
 
-        #checks that one of $slug and $ditemid exists
-        return 0 unless $ditemid || $slug;
+        # This shouldn't happen but, just in case, we check the number of new stickies and
+        # if we have more than we're allowed we trim the input array accordingly.
+        @input = @input[0..$max_sticky_count-1] unless $new_sticky_count < $max_sticky_count;
 
-        # Validate the entry
-        my $item;
+        # We add the currently_unused_stickies onto the end of the new stickies.
+        # This has the side effect that, if the user hasn't allocated all their
+        # sticky quota but has previously used more than their quota that some of their
+        # old stickies will "shuffle up" to fill in the space.
+        my $sticky_entry = join( ',', ( @input, @new_unused_stickies ) );
+        $u->set_prop( sticky_entry => $sticky_entry );
+        return 1;
+    }
 
-        $item = LJ::Entry->new( $u, ditemid => $ditemid ) if $ditemid;
+    my @entries = map { LJ::Entry->new( $u, ditemid => $_ ) } @entry_ids;
+    @entries = @entries[0..$max_sticky_count-1] if scalar @entries > $max_sticky_count;
+    return @entries;
+}
 
-        if ( $slug ) {
-            $item = LJ::Entry->new( $u, slug => $slug );
-            $ditemid = $item->ditemid;
-        }
+# returns a map of ditemid => 1 of the sticky entries
+sub sticky_entries_lookup {
+    my ( $u ) = @_;
 
-        return 0 unless $item && $item->valid;
+    my @stickies = split /,/, $u->prop( 'sticky_entry' );
+    return { map { $_ => 1 } @stickies };
+}
 
-        $u->set_prop( sticky_entry => $ditemid );
-        return 1;
-    }
-    return $u->prop( 'sticky_entry' );
+# Make a particular entry into a particular sticky.
+sub sticky_entry_new {
+    my ( $u, $ditemid ) = @_;
+
+    my $remote = LJ::get_remote();
+    return unless $remote && $remote->can_manage( $u );
+
+    my @stickies = split /,/, $u->prop( 'sticky_entry' );
+    return undef if scalar @stickies >= $u->count_max_stickies;
+
+    unshift @stickies, $ditemid;
+    my $sticky_entry_list = join( ',', @stickies );
+
+    $u->set_prop( sticky_entry => $sticky_entry_list );
+
+    return 1;
 }
 
-sub get_sticky_entry {
-    my $u = shift;
+# Remove a particular entry from the sticky list
+sub sticky_entry_remove {
+    my ( $u, $ditemid ) = @_;
 
-    if ( my $ditemid = $u->sticky_entry ) {
-        my $item = LJ::Entry->new( $u, ditemid => $ditemid );
-        return $item if $item->valid;
-    }
-    return undef;
+    my $remote = LJ::get_remote();
+    return unless $remote && $remote->can_manage( $u );
+
+    my $sticky_entry_ids =  $u->prop( 'sticky_entry' );
+    my @stickies = split( /,/, $sticky_entry_ids );
```

Comment:
I don't think that sticky_entries_lookup will work here because that won't maintain the order, but I'm taking the hint and making a convenience method ;)